### PR TITLE
rust 1.61.0

### PR DESCRIPTION
**Pull request by Porkepix**
Created with `brew bump-formula-pr`.

~`resource` blocks may require updates.~

### CODE REVIEW

File: Formula/rust.rb [hunk 1]
```diff
@@ -4,8 +4,8 @@ class Rust < Formula
   license any_of: ["Apache-2.0", "MIT"]
 
   stable do
-    url "https://static.rust-lang.org/dist/rustc-1.59.0-src.tar.gz"
-    sha256 "a7c8eeaee85bfcef84c96b02b3171d1e6540d15179ff83dddd9eafba185f85f9"
+    url "https://static.rust-lang.org/dist/rustc-1.60.0-src.tar.gz"
+    sha256 "20ca826d1cf674daf8e22c4f8c4b9743af07973211c839b85839742314c838b7"
 
     # From https://github.com/rust-lang/rust/tree/#{version}/src/tools
     resource "cargo" do
```
Comment (Bo98) [inline Formula/rust.rb:11]: This needs updating to 0.61.0

**Comment by dtrodrigues** [comment]
Can this build with python@3.10?

**Comment by EricFromCanada** [comment]
Also on 12-Intel. Progress, I guess? (I don't seen any errors when compiling locally on 12.3.)
```
     Compiling openssl-sys v0.9.72
...
  error: could not compile `openssl-sys`
  
  Caused by:
    process didn't exit successfully: `/usr/local/Cellar/rust/1.60.0/bin/rustc --crate-name openssl_sys /Users/brew/Library/Caches/Homebrew/cargo_cache/registry/src/github.com-1ecc6299db9ec823/openssl-sys-0.9.72/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no -C metadata=008c9680d21c68a1 -C extra-filename=-008c9680d21c68a1 --out-dir /private/tmp/rust--cargo-20220414-46316-k28i0w/target/release/deps -L dependency=/private/tmp/rust--cargo-20220414-46316-k28i0w/target/release/deps --extern libc=/private/tmp/rust--cargo-20220414-46316-k28i0w/target/release/deps/liblibc-fa93ba606703e75c.rmeta --cap-lints allow -L 'native=/usr/local/opt/openssl@1.1/lib' --cfg const_fn --cfg 'osslconf="OPENSSL_NO_SSL3_METHOD"' --cfg ossl101 --cfg ossl102 --cfg ossl102f --cfg ossl102h --cfg ossl110 --cfg ossl110f --cfg ossl110g --cfg ossl110h --cfg ossl111 --cfg ossl111b --cfg ossl111c -l dylib=ssl -l dylib=crypto` (signal: 6, SIGABRT: process abort signal)
```

**Comment by dtrodrigues** [comment]
On arm:
```
Warning: 2 failed steps ignored!
brew install --build-from-source rustscan
brew install --build-from-source sslyze
Error: 1 failed step!
brew install --build-from-source magic-wormhole
```
For rustscan, it should work once https://github.com/RustScan/RustScan/pull/431 is merged.

For magic-wormhole, the build failure seems unrelated to the new version, and should be fixed by https://github.com/Homebrew/homebrew-core/pull/99428

**Comment by cho-m** [comment]
Intel Monterey keeps failing. Now back to first error
```
     Compiling combine v4.6.3
     Compiling core-foundation v0.9.3
     Compiling bitmaps v2.1.0
     Compiling tar v0.4.38
  error: could not compile `combine`
  
  Caused by:
    process didn't exit successfully: `/usr/local/Cellar/rust/1.60.0/bin/rustc --crate-name combine --edition=2018 /Users/brew/Library/Caches/Homebrew/cargo_cache/registry/src/github.com-1ecc6299db9ec823/combine-4.6.3/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --cfg 'feature="alloc"' --cfg 'feature="bytes"' --cfg 'feature="default"' --cfg 'feature="std"' -C metadata=946663382fe94ea6 -C extra-filename=-946663382fe94ea6 --out-dir /private/tmp/rust--cargo-20220418-88264-1mrk4np/target/release/deps -L dependency=/private/tmp/rust--cargo-20220418-88264-1mrk4np/target/release/deps --extern bytes=/private/tmp/rust--cargo-20220418-88264-1mrk4np/target/release/deps/libbytes-8cb23b4685950048.rmeta --extern memchr=/private/tmp/rust--cargo-20220418-88264-1mrk4np/target/release/deps/libmemchr-a49f41a35aac3a66.rmeta --cap-lints allow` (signal: 6, SIGABRT: process abort signal)
  warning: build failed, waiting for other jobs to finish...
  error: failed to compile `cargo v0.61.0 (/private/tmp/rust--cargo-20220418-88264-1mrk4np)`, intermediate artifacts can be found at `/private/tmp/rust--cargo-20220418-88264-1mrk4np/target`
```

**Comment by Bo98** [comment]
Looks like a genuine crash bug in the compiler, potentially introduced in this new version. I'll poke around for crash logs.

**Comment by chenrui333** [comment]
> Looks like a genuine crash bug in the compiler, potentially introduced in this new version. I'll poke around for crash logs.

did we also report into the upstream for better visibility?

**Comment by Bo98** [comment]
It's already reported: https://github.com/rust-lang/rust/issues/92173.

Enabling jemalloc should fix the issue (not sure if that means using a bundled one or if we can use ours or what).

**Comment by cho-m** [comment]
I've went ahead and added system `jemalloc`. Not sure whether we want that or not though based on Rust's design philosophy, but may as well try it out.
```
Homebrew libraries:
  /opt/homebrew/opt/jemalloc/lib/libjemalloc.2.dylib (jemalloc)
```

**Comment by Bo98** [comment]
Looks like Rust's jemalloc is just upstream + a couple backported patches that came later on (but both are profiling related which I don't think affects rustc) so we should indeed use ours.

**Comment by cho-m** [comment]
Catalina hit 3 day limit.
May want to first make sure that current run looks good, then re-run without building dependents.

**Comment by cho-m** [comment]
I guess one issue of adding `jemalloc` is increased build time, which may cause future `rust` PRs to time out during dependent testing.  Also, may want to analyze impact in binary size.

Intel Big Sur went from 2days 6hours to 2days 21hours (+15h) while Catalina went from 2days 10hours to 3days+. This would cause some unwanted overhead if run times out after 3 days.

**Comment by toffaletti** [comment]
> I guess one issue of adding `jemalloc` is increased build time

probably because you're linking a shared jemalloc instead of statically linking it which allows for more optimization opportunities

**Comment by samford** [comment]
Just a heads up: Rust 1.61.0 has been released, whenever this is ready for another run.

File: Formula/rust.rb [hunk 2]
```diff
@@ -4,14 +4,14 @@ class Rust < Formula
   license any_of: ["Apache-2.0", "MIT"]
 
   stable do
-    url "https://static.rust-lang.org/dist/rustc-1.59.0-src.tar.gz"
-    sha256 "a7c8eeaee85bfcef84c96b02b3171d1e6540d15179ff83dddd9eafba185f85f9"
+    url "https://static.rust-lang.org/dist/rustc-1.60.0-src.tar.gz"
+    sha256 "20ca826d1cf674daf8e22c4f8c4b9743af07973211c839b85839742314c838b7"
```
Comment (Barbiero) [inline Formula/rust.rb]:
```suggestion
    url "https://static.rust-lang.org/dist/rustc-1.61.0-src.tar.gz"
    sha256 "ad0b4351675aa9abdf4c7e066613bd274c4391c5506db152983426376101daed"
```

File: Formula/rust.rb [hunk 3]
```diff
@@ -4,14 +4,14 @@ class Rust < Formula
...
 
     # From https://github.com/rust-lang/rust/tree/#{version}/src/tools
     resource "cargo" do
       url "https://github.com/rust-lang/cargo.git",
-          tag:      "0.60.0",
-          revision: "49d8809dc2d3e6e0d5ec634fcf26d8e2aab67130"
+          tag:      "0.61.0",
+          revision: "d1fd9fe2c40a1a56af9132b5c92ab963ac7ae422"
```
Comment (Barbiero) [inline Formula/rust.rb]:
```suggestion
          tag:      "0.62.0",
          revision: "a028ae42fc1376571de836be702e840ca8e060c2"
```

File: Formula/rust.rb [hunk 4]
```diff
@@ -46,29 +47,39 @@ class Rust < Formula
     on_macos do
       # From https://github.com/rust-lang/rust/blob/#{version}/src/stage0.json
       if Hardware::CPU.arm?
-        url "https://static.rust-lang.org/dist/2022-01-13/cargo-1.58.0-aarch64-apple-darwin.tar.gz"
-        sha256 "9144ee0f614c8dcb5f34a774e47a24b676860fa442afda2a3c7f45abfe694e6a"
+        url "https://static.rust-lang.org/dist/2022-02-24/cargo-1.59.0-aarch64-apple-darwin.tar.gz"
+        sha256 "4bdcb77ea084364b551a5cf969e263beb09afa39627f6dead262c8e2a7aed9c1"
```
Comment (Barbiero) [inline Formula/rust.rb]:
```suggestion
        url "https://static.rust-lang.org/dist/2022-05-19/cargo-1.61.0-aarch64-apple-darwin.tar.gz"
        sha256 "8099a35548e1ae4773dbbcbe797301c500ec10236435fde0073f52b4937be6c3"
```

File: Formula/rust.rb [hunk 5]
```diff
@@ -46,29 +47,39 @@ class Rust < Formula
...
       else
-        url "https://static.rust-lang.org/dist/2022-01-13/cargo-1.58.0-x86_64-apple-darwin.tar.gz"
-        sha256 "60203fc7ec453f2a9eb93734c70a72f8ee88e349905edded04155c1646e283a6"
+        url "https://static.rust-lang.org/dist/2022-02-24/cargo-1.59.0-x86_64-apple-darwin.tar.gz"
+        sha256 "d0202b50b4f3d0e943a5d7e8d14420afeac8cf36e6136e8d5b7ddefb3538d62a"
```
Comment (Barbiero) [inline Formula/rust.rb]:
```suggestion
        url "https://static.rust-lang.org/dist/2022-05-19/cargo-1.61.0-x86_64-apple-darwin.tar.gz"
        sha256 "41154ccc20ade7b14a2f29ba1b7777684b87495ae94570de6726ae72731acc95"
```

File: Formula/rust.rb [hunk 6]
```diff
@@ -46,29 +47,39 @@ class Rust < Formula
...
       end
     end
 
     on_linux do
       # From: https://github.com/rust-lang/rust/blob/#{version}/src/stage0.json
-      url "https://static.rust-lang.org/dist/2022-01-13/cargo-1.58.0-x86_64-unknown-linux-gnu.tar.gz"
-      sha256 "940aa91ad2de39c18749e8d789d88846de2debbcf6207247225b42c6c3bf731a"
+      url "https://static.rust-lang.org/dist/2022-02-24/cargo-1.59.0-x86_64-unknown-linux-gnu.tar.gz"
+      sha256 "703c1c66f1d68549d1ee0b7a96f91935a3583af8ec13cc1722ff7a0f1c69ca70"
```
Comment (Barbiero) [inline Formula/rust.rb]:
```suggestion
      url "https://static.rust-lang.org/dist/2022-05-19/cargo-1.61.0-x86_64-unknown-linux-gnu.tar.gz"
      sha256 "c6e108e13ef5e08e71d70685861590f8683090368cab1f4eacfe97677333b2c7"
```

**Review by Barbiero — COMMENTED** [review]
added suggestions to update the crate straight to 1.61 / cargo 0.62

EDIT: well that's not fair, @branchvincent commited these literally a minute before I added the suggestions! 😄

**Comment by cho-m** [comment]
> probably because you're linking a shared jemalloc instead of statically linking it which allows for more optimization opportunities

May be worth comparing impact of just statically compiling upstream's `jemalloc` submodule. ArchLinux (https://github.com/archlinux/svntogit-packages/blob/packages/rust/trunk/PKGBUILD#L93) and MacPorts currently build this way.

Otherwise, we will probably hit 3 day limit for all future `rust` PRs until we add support for testing dependents on separate runners (e.g. https://github.com/Homebrew/homebrew-test-bot/issues/727)

**Comment by branchvincent** [comment]
For 11/12 ARM failures:

* `zoxide`: checksum mismatch 
* all others: a `pip` bug throwing `'extra' does not exist in evaluation environment.` (fix in  #102092)

**Comment by alex** [comment]
The `python@3.9` changes were merged.

**Comment by hustcer** [comment]
Any update on this?

**Comment by branchvincent** [comment]
> May be worth comparing impact of just statically compiling upstream's `jemalloc`

Just pushed, I think we can try another run. Only `watchman` didn't get addressed

**Comment by hustcer** [comment]
No good luck again

**Comment by cho-m** [comment]
Run probably hit a CI or GitHub issue given that the action looks to be in an odd state. Not sure why, but run got cancelled. Will need to re-run later, though will have to wait for another free long-timeout slot.
```
2022-06-16T12:20:29.9745830Z ==> brew install --build-from-source --only-dependencies aws-sso-util
2022-06-16T12:20:33.2905730Z ==> brew install --build-from-source aws-sso-util
2022-06-16T12:41:01.7964770Z ##[error]The operation was canceled.
2022-06-16T12:41:01.8111560Z ##[group]Run touch bottles/steps_output.txt
```

**Comment by cho-m** [comment]
`11-arm64` had sporadic build failure. We can re-run by itself once everything else finishes.

`12-arm64` failures seem unrelated:
- `pwntools` - sporadic network failure
- `cjdns` - looks like it is due to Monterey `python` removal
- `rgbds` - not sure yet, but can reproduce test failure when locally building from source with `rust 1.59.0`. Existing bottles work, so something seems to have changed in build environment after previous bottling and prior to this PR.

Intel macOS runners at least haven't been randomly cancelled. `12-arm64` went down from 16h to 13h run, so some hope that it doesn't hit 3 day limit. Probably still need to wait 1 more day to get results.

**Comment by cho-m** [comment]
Pretty much all Intel macOS finished in 2d 10h with `rgbds` + `watchman` failures. Looking at last PRs for these, `rgbds` takes ~7m while `watchman` takes ~35m so should be okay on total CI time.

Only other failure is `watchman` which should be checked to see if unrelated. It looks like `watchman` built/passed on `12-arm64` so there is good chance it is something else that is causing failure on Intel.

Kicked off a re-run of `11-arm64` and should finish in half a day or so. Once CI is done and we finish looking into `watchman`, then PR should be good to merge.

**Comment by cho-m** [comment]
I think `watchman` may be failing since when we built `folly`/`watchman` we used march (e.g. `nehalem`) for compatibility; however, we run dependent tests with `--build-from-source` which doesn't seem to set any march.

`folly` doesn't like mixing different SSE/etc flags https://github.com/facebook/folly/blob/v2022.06.13.00/folly/container/detail/F14Table.cpp#L26-L31
```cpp
// If you get a link failure that leads you here, your build has varying
// compiler flags across compilation units in a way that would break F14.
// SIMD (SSE2 or NEON) needs to be either on everywhere or off everywhere
// that uses F14.  If SIMD is on then hardware CRC needs to be enabled
// everywhere or disabled everywhere.
void F14LinkCheck<getF14IntrinsicsMode()>::check() noexcept {}
```

So, I think it is safe to ignore. We can triage any other issues when we bump to `2022.06.20.00`.

**Review by cho-m — APPROVED** [review]

**Comment by chenrui333** [comment]
Thanks @cho-m!!

**Pull request closed without merging**